src/inmanta/data/__init__.py

```diff
@@ -2049,6 +2049,32 @@ class Resource(BaseDocument):
     # the list contains full rv id's
     provides: List[m.ResourceVersionIdStr] = Field(field_type=list, default=[])  # List of resource versions
 
+    @classmethod
+    async def get_resource_state_for_dependencies(
+        cls, environment: uuid.UUID, resource_version_id: resources.Id
+    ) -> Dict[m.ResourceVersionIdStr, ResourceState]:
+        """
+        Return the ResourceState for each dependency of the given resource.
+        """
+        if not resource_version_id.is_resource_version_id_obj():
+            raise Exception("Argument resource_version_id is not a resource_version_id")
+        query = f"""
+            SELECT r1.resource_version_id, r1.status
+            FROM {cls.table_name()} AS r1
+            WHERE r1.environment=$1 AND r1.model=$2 AND (
+                                                            SELECT (r2.attributes->'requires')::jsonb
+                                                            FROM {cls.table_name()} AS r2
+                                                            WHERE r2.environment=$1 AND r2.resource_version_id=$3
+                                                         ) ? r1.resource_version_id
```

Comment:
wow, never seen this before.

if I understand correctly, this is the jsonb containment operator?

(I would love to see the query plan for this beasts ;-) )

Reply:
Yes. It checks whether an element is present in the list `attributes->require`, 

I extracted a query plan from mgmt in order to get realistic output:

```
 Index Scan using resource_env_model_agent_index on resource r1  (cost=8.85..6180.59 rows=3 width=86) (actual time=0.269..2.778 rows=2 loops=1)
   Index Cond: ((environment = 'f81be80b-8582-499f-b002-7cdd2f2df98a'::uuid) AND (model = 1572536771))
   Filter: ($0 ? (resource_version_id)::text)
   Rows Removed by Filter: 3061
   InitPlan 1 (returns $0)
     ->  Index Scan using resource_pkey on resource r2  (cost=0.41..8.44 rows=1 width=32) (actual time=0.024..0.025 rows=1 loops=1)
           Index Cond: ((environment = 'f81be80b-8582-499f-b002-7cdd2f2df98a'::uuid) AND ((resource_version_id)::text = 'std::Service[artifacts,name=yum-cron],v=1572536771'::text))
 Planning time: 0.135 ms
 Execution time: 2.797 ms
```